vindex/cmd/client/client.go

```diff
@@ -183,3 +216,85 @@ func (c VIndexClient) lookupUnverified(ctx context.Context, key string) (api.Loo
 	}
 	return lookupResp, nil
 }
+
+func NewLeafReaderFromFlags(ctx context.Context) (*LeafReader, error) {
+	inV, err := note.NewVerifier(*inLogPubKey)
+	if err != nil {
+		klog.Exitf("failed to construct input log verifier: %v", err)
+	}
+
+	if *inLogBaseURL == "" {
+		klog.Exit("in_log_base_url flag must be provided")
+	}
+	u, err := url.Parse(*inLogBaseURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse URL: %v", err)
+	}
+	c, err := client.NewHTTPFetcher(u, http.DefaultClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create HTTP fetcher for %q: %v", u, err)
+	}
+
+	cpRaw, err := c.ReadCheckpoint(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read checkpoint from input log: %v", err)
+	}
+	cp, _, _, err := log.ParseCheckpoint(cpRaw, inV.Name(), inV)
+	if err != nil {
+		klog.Warning(string(cpRaw))
+		return nil, fmt.Errorf("failed to parse input log checkpoint: %v", err)
+	}
+	lr := &LeafReader{
+		f:  c.ReadEntryBundle,
+		cp: cp,
+	}
+	return lr, nil
+}
+
+// LeafReader reads leaves from the tree.
+// This class is not thread safe.
+type LeafReader struct {
+	f  client.EntryBundleFetcherFunc
+	cp *log.Checkpoint
+	c  leafBundleCache
+}
+
+// getLeaf fetches the raw contents committed to at a given leaf index.
+func (r *LeafReader) getLeaf(ctx context.Context, i uint64) ([]byte, error) {
+	if i >= r.cp.Size {
+		return nil, fmt.Errorf("requested leaf %d >= log size %d", i, r.cp.Size)
+	}
+	if cached, _ := r.c.get(i); cached != nil {
+		klog.V(2).Infof("Using cached result for index %d", i)
+		return cached, nil
+	}
+
+	// TODO(mhutchinson): Check the inclusion proof of a fetched bundle
+	bundle, err := client.GetEntryBundle(ctx, r.f, i/layout.EntryBundleWidth, r.cp.Size)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get entry bundle: %v", err)
+	}
+	ti := i % layout.EntryBundleWidth
+	r.c = leafBundleCache{
+		start:  i - ti,
+		leaves: bundle.Entries,
+	}
+	return r.c.leaves[ti], nil
+}
+
+// leafBundleCache stores the results of the last fetched tile. Assuming that the client
+// accesses leaves in order, then this avoids fetching the same bundle multiple times if
+// multiple leaves are in the same bundle.
+type leafBundleCache struct {
+	start  uint64
+	leaves [][]byte
+}
+
+func (tc leafBundleCache) get(i uint64) ([]byte, error) {
```

Comment:
You don't use the `error` AFAICT, so you could just drop it and rely on the `[]byte` being `nil`?